plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java

```diff
@@ -1146,6 +1153,217 @@ public void testExecuteProcedureWithInvalidQuery()
+                .isFullyPushedDown();
+        assertThat(query(session, "SELECT * FROM (SELECT * FROM %s WHERE regionkey < 3) LIMIT 2".formatted(tableName)))
+                .skipResultsCorrectnessCheckForPushdown()
+                .isFullyPushedDown();
+
+        // With native query
+        assertNativeQueryWithLimitPushdown(session, tableName);
+        // With Empty projection
+        assertEmptyProjectionWithLimitPushdown(session, tableName);
+    }
+
+    private void assertNativeQueryWithLimitPushdown(Session session, String tableName)
+    {
+        assertThat(query(session, "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT * FROM %s')) LIMIT 0".formatted(tableName)))
+                .returnsEmptyResult();
+        assertThat(query(session, "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT * FROM %s')) LIMIT 3".formatted(tableName)))
+                .skipResultsCorrectnessCheckForPushdown()
+                .isFullyPushedDown();
+        assertThat(query(session, "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT * FROM %s')) LIMIT 10".formatted(tableName)))
+                .isFullyPushedDown();
+
+        assertThat(query(session, "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT * FROM %s')) WHERE regionkey < 3 LIMIT 2".formatted(tableName)))
+                .skipResultsCorrectnessCheckForPushdown()
+                .isFullyPushedDown();
+        assertThat(query(session, "SELECT * FROM TABLE(bigquery.system.query(query => 'SELECT * FROM %s')) WHERE regionkey > 5 LIMIT 2".formatted(tableName)))
+                .returnsEmptyResult();
+    }
+
+    private void assertEmptyProjectionWithLimitPushdown(Session session, String tableName)
+    {
+        PlanMatchPattern aggregationOverTableScan = node(AggregationNode.class, anyTree(node(TableScanNode.class)));
+
+        assertThat(query(session, "SELECT COUNT(*) FROM %s LIMIT 0".formatted(tableName)))
+                .returnsEmptyResult();
+        assertThat(query(session, "SELECT COUNT(*) FROM %s LIMIT 1".formatted(tableName)))
+                .matches("VALUES BIGINT '5'")
+                .isNotFullyPushedDown(aggregationOverTableScan);
+        assertThat(query(session, "SELECT COUNT(*) FROM %s LIMIT 10".formatted(tableName)))
+                .matches("VALUES BIGINT '5'")
+                .isNotFullyPushedDown(aggregationOverTableScan);
+
+        assertThat(query(session, "SELECT COUNT(*) FROM (SELECT * FROM %s LIMIT 0)".formatted(tableName)))
+                .matches("VALUES BIGINT '0'");
+        assertThat(query(session, "SELECT COUNT(*) FROM (SELECT * FROM %s LIMIT 3)".formatted(tableName)))
+                .matches("VALUES BIGINT '3'")
+                .isNotFullyPushedDown(aggregationOverTableScan);
+        assertThat(query(session, "SELECT COUNT(*) FROM (SELECT * FROM %s LIMIT 10)".formatted(tableName)))
+                .matches("VALUES BIGINT '5'")
+                .isNotFullyPushedDown(aggregationOverTableScan);
```

Comment:
How do we ensure that `LIMIT` is pushed down here ? Can we add `anyNot(LimitNode.class, ...)` to ensure that Limit pushdown did happen here.

Reply:
As discussed offline, Using `tableScan` to verify limit value.

Reply:
Now verifying `tableScan` with limit along with `anyNot(LimitNode.class, ...)`

```
anyNot(
    LimitNode.class, node(
            AggregationNode.class, anyTree(node(
                    AggregationNode.class,
                    tableScan(table -> {
                        BigQueryTableHandle actualTableHandle = (BigQueryTableHandle) table;
                        return actualTableHandle.limit().equals(limit);
                    },
                    TupleDomain.all(),
                    ImmutableMap.of())))));
```

Reply:
CC: @Praveen2112 ^^